In [1]:
#importing pandas
import pandas as pd

In [2]:
#steps in ML
# 1. Loading dataset
# 2. Preprocessing dataset (handling missing values etc)
# 3. Dividing dataset into training and testing
# 4. Training ML model with train dataset
# 5. Testing the result 
# 6  Predicting for testing dataset

In [6]:
#loading dataset
df=pd.read_csv("titanic.csv")

In [7]:
df.head()

PassengerId  Survived  Pclass  ... Cabin Embarked  Child
0            2         1       1  ...   C85        C      0
1            4         1       1  ...  C123        S      0
2            7         0       1  ...   E46        S      0
3           11         1       3  ...    G6        S      1
4           12         1       1  ...  C103        S      0

[5 rows x 13 columns]

In [8]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'Child'],
      dtype='object')

In [11]:
#after removing extra columns
df=df[["Survived","Pclass","Sex","Age","SibSp","Embarked","Child"]]

In [12]:
df.head()

Survived  Pclass     Sex   Age  SibSp Embarked  Child
0         1       1  female  38.0      1        C      0
1         1       1  female  35.0      1        S      0
2         0       1    male  54.0      0        S      0
3         1       3  female   4.0      1        S      1
4         1       1  female  58.0      0        S      0

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183 entries, 0 to 182
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  183 non-null    int64  
 1   Pclass    183 non-null    int64  
 2   Sex       183 non-null    object 
 3   Age       183 non-null    float64
 4   SibSp     183 non-null    int64  
 5   Embarked  183 non-null    object 
 6   Child     183 non-null    int64  
dtypes: float64(1), int64(4), object(2)
memory usage: 10.1+ KB


In [14]:
#convert object into numeric data- Compulsory for all columns to be in numeric form
#Label encoder- It is used to create labels in the columns
#Sklearn library- The library for traditional ML and preproceeing such as Encoding, Normalisation, Standardisation

In [17]:
from sklearn import preprocessing

In [18]:
#le is instance here for the below function 
le=preprocessing.LabelEncoder()

In [20]:
#encoding column sex and embarked
df["Sex"]=le.fit_transform(df["Sex"])
df["Embarked"]=le.fit_transform(df["Embarked"])

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [21]:
df.head()

Survived  Pclass  Sex   Age  SibSp  Embarked  Child
0         1       1    0  38.0      1         0      0
1         1       1    0  35.0      1         2      0
2         0       1    1  54.0      0         2      0
3         1       3    0   4.0      1         2      1
4         1       1    0  58.0      0         2      0

0 for female, 1 for male 

C=0, S=2

In [22]:
# dataset into input and output variables
# X for input y for output
X=df.drop("Survived",axis=1)  #axis 1 for column and 0 for row
y=df["Survived"]

In [23]:
# Splitting dataset
# Train data- for traning the model || Test data- for evaluating the model

In [25]:
from sklearn.model_selection import train_test_split

In [57]:
#train_test_split will convert our data to train set and test set
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.2)
#here 80% data is training dta and 20% data of dataset is test data

X_train,y_train will train our model

X_test,y_test will test our model

In [58]:
#KNN

KNN- K nearest neighbours 
so basically this algo calculates distance of an unknown point (that needs to be predicted) from the k nearest neighbours of the two groups independently and will assign the point to the nearest group found.

It is also known as lazy learning algo as it needs the point that needs to be predicted

Simplest classifier model

In [59]:
from sklearn.neighbors import KNeighborsClassifier

In [60]:
# training dataset
model=KNeighborsClassifier(n_neighbors=3)   #let k=3 here
model.fit(X_train, y_train)   #model_fit: for training model

KNeighborsClassifier(n_neighbors=3)

In [61]:
X_train.shape

(146, 6)

In [62]:
y_train.shape

(146,)

In [63]:
# Testing the KNN model
y_predicted=model.predict(X_test)   #mode.predict: for predicting using trained model

In [64]:
# Evaluation of model

We will caculate the accuracy score agains y_test

y_test is the ground truth value || 
y_predicted is the predicted value we got from model

In [65]:
from sklearn.metrics import accuracy_score

In [66]:
# syntax of accuracy_score is accuracy_score(predicted,ground value)
accuracy_score(y_predicted,y_test)*100

64.86486486486487

The accurracy found here from KNN model (k=3) is 65% approx.